src/NemoStuff.jl

```diff
@@ -0,0 +1,804 @@
+    w, n = remove(numerator(z), p)
+    return w - v, n // d
+end
+
+function valuation(z::Rational{T}, p::T) where {T<:Integer}
+    z == 0 && error("Not yet implemented")
+    v = valuation(denominator(z), p)
+    w = valuation(numerator(z), p)
+    return w - v
+end
+
+function matrix(a::Vector{Vector{T}}) where {T}
+    return matrix(permutedims(reduce(hcat, a), (2, 1)))
+end
+
+################################################################################
+#
+#  Create a matrix from rows
+#
+################################################################################
+
+function matrix(K::Ring, R::Vector{<:Vector})
+    if length(R) == 0
+        return zero_matrix(K, 0, 0)
+    else
+        n = length(R)
+        m = length(R[1])
+        z = zero_matrix(K, n, m)
+        for i in 1:n
+            @assert length(R[i]) == m
+            for j in 1:m
+                z[i, j] = R[i][j]
+            end
+        end
+        return z
+    end
+end
+
+################################################################################
+#
+#  Diagonal (block) matrix creation
+#
+################################################################################
+
+@doc raw"""
+    diagonal_matrix(x::T...) where T <: RingElement -> MatElem{T}
+    diagonal_matrix(x::Vector{T}) where T <: RingElement -> MatElem{T}
+    diagonal_matrix(Q, x::Vector{T}) where T <: RingElement -> MatElem{T}
+
+Returns a diagonal matrix whose diagonal entries are the elements of $x$.
+
+# Examples
+
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> diagonal_matrix(ZZ(1), ZZ(2))
+[1   0]
+[0   2]
+
+julia> diagonal_matrix([ZZ(3), ZZ(4)])
+[3   0]
+[0   4]
+
+julia> diagonal_matrix(ZZ, [5, 6])
+[5   0]
+[0   6]
+```
+"""
+function diagonal_matrix(R::Ring, x::Vector{<:RingElement})
+    x = R.(x)
+    M = zero_matrix(R, length(x), length(x))
+    for i = 1:length(x)
+        M[i, i] = x[i]
+    end
+    return M
+end
+
+function diagonal_matrix(x::T, xs::T...) where {T<:RingElement}
+    return diagonal_matrix(collect((x, xs...)))
+end
+
+diagonal_matrix(x::Vector{<:RingElement}) = diagonal_matrix(parent(x[1]), x)
+
+@doc raw"""
+    diagonal_matrix(x::Vector{T}) where T <: MatElem -> MatElem
+
+Returns a block diagonal matrix whose diagonal blocks are the matrices in $x$.
+"""
+function diagonal_matrix(x::Vector{T}) where {T<:MatElem}
+    return cat(x..., dims=(1, 2))::T
+end
+
+function diagonal_matrix(x::T, xs::T...) where {T<:MatElem}
+    return cat(x, xs..., dims=(1, 2))::T
+end
+
+function diagonal_matrix(R::Ring, x::Vector{<:MatElem})
+    if length(x) == 0
+        return zero_matrix(R, 0, 0)
+    end
+    x = [change_base_ring(R, i) for i in x]
+    return diagonal_matrix(x)
+end
+
+base_ring(::Vector{Int}) = Int
+
+function is_symmetric(M::MatElem)
+    for i in 1:nrows(M)
+        for j in i:ncols(M)
+            if M[i, j] != M[j, i]
+                return false
+            end
+        end
+    end
+    return true
+end
```

Comment:
We already have the following in `src/Matrix.jl:1302` -- these should be merged (probably by just dropping the restriction on `T` on the existing method)
```
function is_symmetric(a::MatrixElem{T}) where T <: NCRingElement
    if !is_square(a)
        return false
    end
    for row in 2:nrows(a)
        for col in 1:(row - 1)
            if a[row, col] != a[col, row]
                return false
            end
        end
    end
    return true
end
```